### DISCUSSION TITLE
autogen api service to swr hooks ideas

### DISCUSSION THREAD

**Discussion by darrenjennings** [vercel/swr#1002] [Ideas]
I am curious if you have ever considered doing something with swr similar to how rtk query creates api services: https://rtk-query-docs.netlify.app/introduction/getting-started/#create-an-api-service

I am trying to imagine how to do something where there is a workflow for swagger/oas spec all the way to the swr hook.

OpenAPI Generator ->
https://github.com/OpenAPITools/openapi-generator#overview

Something like:
```ts
// useVercelApi.ts
import client from 'vercel-api-client'; // some autogen client library, like axios-typescript from openapi-generator

export default useSwrClient({client}); // hand-waving magic swr code
```

```ts
// Users.tsx
import { useUsers } from 'useVercelApi'

function App () {
  const { data, error } = useUsers()
  
  return <div>Total Users: {users?.length}</div>
}
```

**Comment by shuding** [comment]
Love the idea! It can be a plugin of the backend framework indeed, and it will work well in a monolithic codebase. 

Related: https://twitter.com/rauchg/status/1368283797280550912.

**Comment by renet** [comment]
I know this is an older post, but just for everyone who stumbles upon this, there is a project that does exactly this: https://orval.dev/ - Not exclusively for, but including swr. Console output is not perfect - if there is something wrong with the specs, it takes a bit of time to figure out how to fix it, but apart from this, it's working pretty well for me.

**Comment by teobler** [comment]
I have a [tool](https://github.com/teobler/swr-request-generator) for generating openapi swagger doc to swr, used it for several years, can have a try...
